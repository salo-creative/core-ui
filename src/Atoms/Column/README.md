# Column

Column is a pure styled component render. Renders a standard `<div>` with themed styling

-------

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Column from '@salo/core-ui/Atoms/Column';
```

Implement as follows

```javascript
<Column>
  // My content
</Column>
```

### For full prop types and usage see storybook info/knobs