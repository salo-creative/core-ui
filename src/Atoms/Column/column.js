import React from 'react';
import styled from 'styled-components';
import { PropTypes } from 'prop-types';

// HELPERS
import { getBreakpoint } from '../../helpers/breakpoints';

export function calculateSize(size, columns) {
  if (!size) return '';
  return `
    width: ${ ((size / columns) * 100) }%;
    flex-basis: ${ ((size / columns) * 100) }%;
    max-width: ${ ((size / columns) * 100) }%;
  `;
}

const StyledColumn = styled.div`
  display: flex;
  width: 100%;
  flex-direction: ${ props => props.flexDirection };
  flex-grow: ${ props => props.flexGrow };
  flex-shrink: ${ props => props.flexShrink };
  flex-basis: ${ props => props.flexBasis };
  align-items: ${ props => props.alignItems };
  justify-content: ${ props => props.justifyContent };
  padding: ${ props => props.padding };

  /* Default size */
  ${ props => calculateSize(props.default, props.columns) }

  /* Phone size */
  ${ getBreakpoint({ min: 'phone' }) } {
    ${ props => calculateSize(props.phone, props.columns) };
  }

  /* Small size */
  ${ getBreakpoint({ min: 'small' }) } {
    ${ props => { return (`${ calculateSize(props.small, props.columns) }`); } }
  }

  /* Tablet size */
  ${ getBreakpoint({ min: 'tablet' }) } {
    ${ props => { return (`${ calculateSize(props.tablet, props.columns) }`); } }
  }

  /* Medium size */
  ${ getBreakpoint({ min: 'medium' }) } {
    ${ props => { return (`${ calculateSize(props.medium, props.columns) }`); } }
  }

  /* Large size */
  ${ getBreakpoint({ min: 'large' }) } {
    ${ props => { return (`${ calculateSize(props.large, props.columns) }`); } }
  }

  /* X Large size */
  ${ getBreakpoint({ min: 'xLarge' }) } {
    ${ props => calculateSize(props.xLarge, props.columns) }
  }
`;

const Column = ({ children, ...props }) => {
  return <StyledColumn { ...props }>{ children }</StyledColumn>;
};

Column.defaultProps = {
  alignItems: 'flex-start',
  columns: 12,
  default: 12,
  flexBasis: '0%',
  flexDirection: 'column',
  flexGrow: '1',
  flexShrink: '0',
  justifyContent: 'flex-start',
  padding: '10px',
  phone: null,
  small: null,
  tablet: null,
  medium: null,
  large: null,
  xLarge: null
};

Column.propTypes = {
  alignItems: PropTypes.oneOf(['flex-start', 'center', 'flex-end']),
  columns: PropTypes.number,
  default: PropTypes.number,
  flexBasis: PropTypes.string,
  flexDirection: PropTypes.oneOf(['row', 'row-reverse', 'column', 'column-reverse']),
  flexGrow: PropTypes.string,
  flexShrink: PropTypes.string,
  justifyContent: PropTypes.oneOf(['space-between', 'center', 'flex-start', 'flex-end']),
  padding: PropTypes.string,
  phone: PropTypes.number,
  small: PropTypes.number,
  tablet: PropTypes.number,
  medium: PropTypes.number,
  large: PropTypes.number,
  xLarge: PropTypes.number
};

export default Column;