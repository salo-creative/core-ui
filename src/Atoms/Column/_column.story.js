import React from 'react';

// load tests
import { withTests } from '@storybook/addon-jest';
import {
  withKnobs, text, select, number
} from '@storybook/addon-knobs';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { Column } from '../../index';

// README //
import README from './README.md';

// Start of story logic
export const Basic = () => {
  const alignItems = select(
    'Align items',
    {
      'Flex start': 'flex-start',
      'Flex end': 'flex-end',
      Center: 'center'
    },
    'flex-start'
  );
  const justifyContent = select(
    'Justify contnet',
    {
      'Space between': 'space-between',
      Center: 'center',
      'Flex start': 'flex-start',
      'Flex end': 'flex-end'
    },
    'flex-start'
  );
  const padding = text('Padding', '10px');
  const columns = number('Columns', 12);
  const defaultWidth = number('Default width', 12);
  const xLarge = number('X Large width (1200px)', 12);
  const large = number('Large width (1000px)', 12);
  const medium = number('Medium Width (860px)', 12);
  const tablet = number('Tablet width (767px)', 12);
  const small = number('Small width (620px)', 12);
  const phone = number('Phone Width (480px)', 12);
  return (
    <Column
      alignItems={ alignItems }
      justifyContent={ justifyContent }
      default={ defaultWidth }
      xLarge={ xLarge }
      large={ large }
      medium={ medium }
      tablet={ tablet }
      small={ small }
      phone={ phone }
      columns={ columns }
      padding={ padding }
      style={ {
        background: '#eaeaea'
      } }
    >
      A lovely Column
    </Column>
  );
};

Basic.story = {
  decorators: [withKnobs, withTests({
    results
  })],
  parameters: {
    notes: README,
    jest: ['column']
  }
};

export default {
  title: 'Atoms/Grid/Column'
};