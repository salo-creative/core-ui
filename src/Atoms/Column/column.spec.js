import React from 'react';
import 'jest-styled-components';

// HELPERS
import { renderWithTheme } from '../../../test';

// COMPONENT
import Column from './index';

test('Render default Column and styles', async () => {
  // Render
  const { getByTestId, container } = renderWithTheme(
    <Column data-testid='test-column'>A tasty treat</Column>
  );
  const el = getByTestId('test-column');
  // Assert
  expect(el).toHaveTextContent('A tasty treat');
  expect(container.firstChild).toMatchSnapshot();
});

test('Render Column and styles with core props', async () => {
  // Render
  const { getByTestId, container } = renderWithTheme(
    <Column
      column={ 16 }
      default={ 10 }
      large={ 10 }
      medium={ 9 }
      tablet={ 8 }
      small={ 7 }
      phone={ 6 }
      padding='20px'
      flexGrow='0'
      flexBasis='10%'
      flexShrink='1'
      flexDirection='row'
      alignItems='center'
      justifyContent='space-between'
      data-testid='test-column'
    >
      A tasty treat
    </Column>
  );
  const el = getByTestId('test-column');
  // Assert
  expect(el).toHaveTextContent('A tasty treat');
  expect(el).toHaveStyleRule('padding', '20px');
  expect(el).toHaveStyleRule('flex-grow', '0');
  expect(el).toHaveStyleRule('flex-shrink', '1');
  expect(el).toHaveStyleRule('flex-direction', 'row');
  expect(el).toHaveStyleRule('align-items', 'center');
  expect(el).toHaveStyleRule('justify-content', 'space-between');
  expect(container.firstChild).toMatchSnapshot();
});