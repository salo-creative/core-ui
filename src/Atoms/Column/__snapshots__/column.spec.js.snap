// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Column and styles with core props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-basis: 10%;
  -ms-flex-preferred-size: 10%;
  flex-basis: 10%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 20px;
  width: 83.33333333333334%;
  -webkit-flex-basis: 83.33333333333334%;
  -ms-flex-preferred-size: 83.33333333333334%;
  flex-basis: 83.33333333333334%;
  max-width: 83.33333333333334%;
}

@media only screen and (min-width:481px) {
  .c0 {
    width: 50%;
    -webkit-flex-basis: 50%;
    -ms-flex-preferred-size: 50%;
    flex-basis: 50%;
    max-width: 50%;
  }
}

@media only screen and (min-width:621px) {
  .c0 {
    width: 58.333333333333336%;
    -webkit-flex-basis: 58.333333333333336%;
    -ms-flex-preferred-size: 58.333333333333336%;
    flex-basis: 58.333333333333336%;
    max-width: 58.333333333333336%;
  }
}

@media only screen and (min-width:781px) {
  .c0 {
    width: 66.66666666666666%;
    -webkit-flex-basis: 66.66666666666666%;
    -ms-flex-preferred-size: 66.66666666666666%;
    flex-basis: 66.66666666666666%;
    max-width: 66.66666666666666%;
  }
}

@media only screen and (min-width:861px) {
  .c0 {
    width: 75%;
    -webkit-flex-basis: 75%;
    -ms-flex-preferred-size: 75%;
    flex-basis: 75%;
    max-width: 75%;
  }
}

@media only screen and (min-width:1001px) {
  .c0 {
    width: 83.33333333333334%;
    -webkit-flex-basis: 83.33333333333334%;
    -ms-flex-preferred-size: 83.33333333333334%;
    flex-basis: 83.33333333333334%;
    max-width: 83.33333333333334%;
  }
}

<div
  class="c0"
  data-testid="test-column"
  default=""
>
  A tasty treat
</div>
`;

exports[`Render default Column and styles 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0%;
  -ms-flex-preferred-size: 0%;
  flex-basis: 0%;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding: 10px;
  width: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  max-width: 100%;
}

<div
  class="c0"
  data-testid="test-column"
  default=""
>
  A tasty treat
</div>
`;
